service/history/handler.go

```diff
@@ -2166,6 +2167,19 @@ func (h *Handler) DeleteDLQTasks(
 	return deletedlqtasks.Invoke(ctx, h.taskQueueManager, request)
 }
 
+// AddTasks calls the [addtasks.Invoke] API with a [shard.Context] for the given shardID. It does this instead of
+// passing in the [shard.Controller] because that's the convention with the other APIs in this handler.
+func (h *Handler) AddTasks(
+	ctx context.Context,
+	request *historyservice.AddTasksRequest,
+) (*historyservice.AddTasksResponse, error) {
+	shardContext, err := h.controller.GetShardByID(request.ShardId)
+	if err != nil {
+		return nil, err
+	}
+	return addtasks.Invoke(ctx, shardContext, h.payloadSerializer, int(h.config.NumberOfShards), request)
```

Comment:
I would follow the pattern of existing APIs and forward the call to historyEngine and then invoke the api logic. From correctness perspective I don't see any difference though, the implementation of shard.AddTasks() internally blocks on historyEngine creation as well, so there's should be no difference. Forwarding to history engine here will make the blocking logic more explicit and not depend on shard context impl.

The lifecycle of shard context and history engine are slightly different. On history engine creation, it will also set some states in the shard context. I think the idea of the code structuure is that only history engine is supposed to access shard context methods. I actually hope we can merge them into one componet some day. I tried before but it's not trivial. :(

Reply:
Done.